### ISSUE TITLE
Toolchanges

### ISSUE THREAD

**Issue by petervanderwalt** [OpenBuilds/OpenBuilds-CONTROL#26]
Normally you'd want the GCODE to have M6 (Pause) then a Tx command.  The UI should pause streaming at the M6, and prompt you to insert tool X.   Then you jog to zero tool length, setzero z, and unpause.   But thats not in the code yet... Sorry!  (: - couple weeks though! 

So the hack here is to split your gcode into seperate files for each tool. The machine will hold position between jobs (unless your gcode resets coordinates with G10, etc).  Run the job file for tool 1.   When its finished, change tool, zero z (dont rezero xy).   Load file for tool 2, go.  etc.    I know its a kludge, but bare with us (; theres a lot of features that still has to be added - and we'll get to them all eventually (:

**Comment by gmorse7** [comment]
Has anything been done to add the capability to accommodate tool changes and resetting Z manually in a single g-code file?

**Comment by openbuilds-engineer** [comment]
Not yet (: 
(Or this issue wouldnt have been "open" right)

**Comment by openbuilds-engineer** [comment]
Working on Toolchanges (finally!) this week:

First light: 

![dropdown](https://user-images.githubusercontent.com/31449271/64199004-ab392080-ce89-11e9-8634-b61b5226553d.png)

**Comment by openbuilds-engineer** [comment]
@swarfer - (: if you have any comments on it so far (and have some more sample gcode files to throw in the testing pool)

**Comment by gmorse7** [comment]
Here's a simple sample with tool change generated by the Fusion 360 GRBL post processor.  I agree, pause on M6, allow movement to facilitate tool change and Z re-zero then resume.  A message box prompt would be nice but given the multitude of possible platforms...unless of course there is a simple way.
[ToolChangeTestPiece.docx](https://github.com/OpenBuilds/OpenBuilds-CONTROL/files/3571388/ToolChangeTestPiece.docx)

**Comment by gmorse7** [comment]
Here's some psuedo code I used to modify a post processor for another machine.

![image](https://user-images.githubusercontent.com/50146581/64209156-a1e19100-ce54-11e9-840c-802d9f95ef4d.png)

**Comment by openbuilds-engineer** [comment]
Ps. Official Gcode spec calls for Tx before M6 (; and can be called many
lines earlier too to give ATCs time to work

**Comment by sharmstr** [comment]
> @swarfer @sharmstr - (: if you have any comments on it so far (and have some more sample gcode files to throw in the testing pool)
> 
> Basically the parser looks at the gcode, finds a M6 and splits it at that point. We look around for related comments to populate the description a little better
> 

As a temp solution, its probably fine.  I know I'm going to press the wrong run button, but you cant code for that :)

I'll get you some sample files tomorrow.

**Comment by swarfer** [comment]
will test it with sketchucam files

**Comment by swarfer** [comment]
Note the Sketchucam can also insert a macro for the toolchange which may not include a M6 command.  User beware (-:

**Comment by openbuilds-engineer** [comment]
Lol (: always right
Does yours have start/end markers we can seperate it out?

**Comment by swarfer** [comment]
yep, just look in the file.  the joiner hooks a few lines from the comment header of each file and includes them in the output just before the tool change command

**Comment by sharmstr** [comment]
Here's one from Fusion using the Tormach post.  I disabled G30 output.

[PegBoard-Fusion-Tormach.txt](https://github.com/OpenBuilds/OpenBuilds-CONTROL/files/3575028/PegBoard-Fusion-Tormach.txt)

**Comment by openbuilds-engineer** [comment]
So did the parser get it right?
Two operations:
1 = Holes cut with a 6.35mm endmil
2 = chamfering with a 12.7mm 30deg 

![two ops](https://user-images.githubusercontent.com/31449271/64269243-dd09c000-cf39-11e9-8b8d-ab052196faec.PNG)

**Comment by openbuilds-engineer** [comment]
re @sharmstr " know I'm going to press the wrong run button, " - the Run Complete Job one is the one I am still working on and the one most should use. Idea is to get it to run to a toolchange, popup a message telling you which tool to load, then lets you jog/probe/etc and finally continue

**Comment by sharmstr** [comment]
The parsed info looks good.

Ah, ok.    During the tool change, will macros be available so I can use a modified G28 for my tool change location instead of having to jog?

**Comment by openbuilds-engineer** [comment]
Yeah I'm forseeing handling each section "like a job" instead of duplicating jog/probe/macro buttons onto a dialog. Just click "ok i'll load up that tool" (hehe) and then we hand you back to the full UI but play button will say "Continue" or something like that

**Comment by openbuilds-engineer** [comment]
@swarfer : 

```
Here is a SketchUcam file with tool changes. This is created by generating a file for T1 and another for T2 and then using the joiner tool to join the files together.
```
Looks like it worked (:

![pockettest](https://user-images.githubusercontent.com/31449271/64274833-7fc73c00-cf44-11e9-8429-07c9669b4f79.PNG)

I checked the file and I do see it having the headers from the joiner as described. 
But for  ```ketchucam can also insert a macro for the toolchange which may not include a M6 command``` - what do you use instead of M6 then
I'm hoping for something like

```
(start of toolchangemacro)
G28
G10 L2 Z0 etc etc
(end of toolchangemacro)
```

So basically I can look for the two comments and know gcode before start is "the job", between the comments is the "toolchange macro" and after end, is the next section with the new tool.  (so the dropdown can reflect accordingly.  When you get to the macro, we either let you manually do it, or run the macro

**Comment by openbuilds-engineer** [comment]
@gmorse7 yours looks good too:

![file2](https://user-images.githubusercontent.com/31449271/64275240-6ecafa80-cf45-11e9-9f08-3fc2fc6dfd27.PNG)

But I dislike how it looks in the 3D viewer (; - so will fix that for you too (;

**Comment by swarfer** [comment]
> 
> 
> @swarfer :
> 
> `Here is a SketchUcam file with tool changes. This is created by generating a file for T1 and another for T2 and then using the joiner tool to join the files together.`
> Looks like it worked (:

yep


> I checked the file and I do see it having the headers from the joiner as described.
> But for `Sketchucam can also insert a macro for the toolchange which may not include a M6 command` - what do you (or more importantly what do I have the parser look for) use instead of M6 then

in this case there is no way to automatically split the file as we have no idea what a user may put into the toolchange macro.   none at all since the macro is user defined.  the point of the macro is to do the tool change without stopping the job so using it and then expecting CONTROL to cooperate with it is a bit of a stretch.

**Comment by openbuilds-engineer** [comment]
If theres no M6, we'll just keep on running. Sounds like that is spot on
then

**Comment by sharmstr** [comment]
How close is this to being released?  Not pressuring, but the fusion post will need to be adjusted.  Just wondering how much time I have :)

**Comment by openbuilds-engineer** [comment]
Phew more than i thought. Had 4 days allocated to work on it but got pulled
off on something else for now. I'd say I am not halfway yet

**Comment by openbuilds-engineer** [comment]
So something like this: Just before the M6, streaming stops, this pops up:

![dialog](https://user-images.githubusercontent.com/31449271/64376056-a8743200-d027-11e9-8eee-e06021696578.PNG)

(also playing with the idea of instead of a dialog popping up, I show this where the 3D viewer usually is) 

You use the regular tools/macros etc to effect the toolchange

then hit the Play button up top (its description now says Continue)

**Comment by sharmstr** [comment]
I like it.  Can you clarify a few things?

1 - I'm assuming the cutter image will always be the same.  If so, I guarantee someone will ask why its not showing their v-bit.  LOL

2 - After rezero'ing, will your continue button raise the cutter?

**Comment by openbuilds-engineer** [comment]
1)  no no - the boss really wants it to show the correct bits or at least close.  It might be "best effort" by looking for keywords like "ball" or "flat" or "v"

2) Hmmm, thats a good Beta tested question (: - i dont think the cam/posts give us a raise before the M6 right, so probably going to have to either do it for users (then i see the "z=5, vs G53 0 fights coming). But excellent catch this will need some debate for sure

**Comment by sharmstr** [comment]
1 - Oh good luck with that.  I can certainly help you with that on the Fusion end.  And maybe even the Aspire end (though I only have the desktop version which doesnt allow multiple tools anyways).

2 - The Tormach sample I gave you does.  Right after the tool change the first move is in Z to your clearance height before any x/y moves.   But it doesnt look like the Autodesk version of grbl post does.  Right after M6 it goes to x/y, then to your Z clearance height.  That could be bad potentially.  I can make sure in our version of the Fusion post that we move to clearance height first if need be, but probably best that control does it.

**Comment by openbuilds-engineer** [comment]
1) If not able to determine, show a cartoonish endmill lol
2) yeah can be really bad, guess we can check what the first move is and decide whether to add one or something. Future Peters problem!  hehe. And yeah we can can fix it in our posts, but you know people are going to load default/other posts etc too

**Comment by gmorse7** [comment]
Does the M6 triggered code raise to safe height and command a spindle stop?  

Tool change user prompt window and text looks good to me.

**Comment by sharmstr** [comment]
That should happen in the gcode before the M6

**Comment by gmorse7** [comment]
Fusion post processors handle tool change coding pretty well from what I've seen.  I've added patches to move to a convenient tool change location in a couple of Fusion posts.  I'm not sure about others.  I see if I can create a couple examples from other CAM software.

**Comment by gmorse7** [comment]
The other attempts I've made to create multi-tool gcode files the software tells me GRBL does not handle tool changes (or at least their post processors don't).

**Comment by swarfer** [comment]
> 
> 
> That should happen in the gcode before the M6.
> 
> Even your ToolChangeTestPiece doc above raises and stops the spindle before the M6 call. _I've never seen a post that wont do that, so I think we are safe there._
> 
> Hmmm, as soon as I say that I now see that Davids file didnt have it. Well, shoot.

yes it does raise Z after the M6
this is the start of the second joined file which uses T2
(File: pockettester.skp)
(Bit diameter: 6.0mm)
(Join   pockettesterT2.nc)
G90 G21 G49 G17 F1500
T2 M06
G53 G00 Z0  ;   <-------------<<    raise Z safely   (-:
G00 X0 Y0

**Comment by sharmstr** [comment]
> (File: pockettester.skp)
> (Bit diameter: 6.0mm)
> (Join pockettesterT2.nc)
> G90 G21 G49 G17 F1500
> T2 M06
> G53 G00 Z0 ; <-------------<< raise Z safely (-:
> G00 X0 Y0

The question was "Does the M6 triggered code raise to safe height and command a **spindle stop**?"

_(File: pockettester.skp)
(Bit diameter: 6.0mm)
(Join   pockettesterT2.nc)
G90 G21 G49 G17 F1500
T2 M06
G53 G00 Z0
G00 X0 Y0
M3 S15000
G00 Z3.000_

It moves up to safety height, but then moves to X0Y0 and sends a spindle start.  Where's the M5??

**Comment by swarfer** [comment]
you are correct! the joiner specifically removes the spindle stops because you really do not want to stop the spindle when joining  several operations with the same tool.  but it leaves the M3 because the new tool may have a different speed.     
I think CONTROL must stop the spindle whenever it finds an M6.  this is the safe thing to do.

**Comment by sharmstr** [comment]
Agreed.  But it should also allow for a spindle stop in the code (okay to be in there, but ignores it).   That way my fusion code can work on other controllers :)

**Comment by gmorse7** [comment]
I agree with sharmstr, the spindle stop should occur immediately after the z retract.

**Comment by openbuilds-engineer** [comment]
I see I am going to need several more hours one this one (; 
Like checking if there is an M5 if not offer one. Etc

**Comment by sharmstr** [comment]
Yeah, sounds like a nightmare.

I havent tested this but I wonder if grbl will throw an error if 2 M5s are issued.  Basically wondering if you actually need to "ignore" an M5 that already in the gcode.  Having to ignore it will be a pain because you cant ignore the last one :)

**Comment by openbuilds-engineer** [comment]
It will accept M5s more than once as it sets a modal. So yeah, good
solution for that particular point

**Comment by petervanderwalt** [comment]
Note to self, once implemented, remember to update Fusion posts etc to support Toolchanges

**Comment by rscamp** [comment]
**Peter**
Do you need any beta testers?  FWIW I have experimented with the bCNC implementation for tool change with the standard GRBL F360 post and it seems to work okay.

**Comment by Kija** [comment]
This looks very promising! Is there any way we can start testing this feature? Has this been implemented in master branch?

**Comment by diegocolonnello** [comment]
Waiting for this to happen, so happy with Openbuilds so far...

**Comment by SeanDolan** [comment]
If I want to do a tool change in the same file, I enter a manual NC I’m Fusion to move the tool head out of the way then another entry for Stop. Measure the tool bit height off the table. Replace the bot and check the height is the same. OpenBuilds Control allows me to hit Pause then Resume and it continues the job. It might not be a resolution outright, but this might help some.

**Comment by sharmstr** [comment]
> Measure the tool bit height off the table. 

How are you taking this measurement?

**Comment by jacobgad** [comment]
I would love to see more attention on this feature i think it makes the whole proses of machining a part much more manageable. thank you for all your hard work

**Comment by southsideboats** [comment]
I wonder was this feature abandoned altogether. If I were smarter I would try and figure it out... it's one thing that makes a Shapeoko so much more user friendly.

**Comment by petervanderwalt** [comment]
Not abandoned no, just haven't been a priority use case as our Post Processors are optimised to split the files for you: See https://docs.openbuilds.com/doku.php?id=docs:software:cam 

In between files you have full control over jogging/zeroing/probing

**Comment by cecheverria** [comment]
Hi guys,

I know OpenSource is hard to maintain, but after 3 months of research, I've almost decided to change My OpenBuilds CNC BlackBox for Acorn CNC or StepCraft machine just for the mere fact of supporting ATC with not much pain. And I'm being really honest, I prefer to spend money on you guys if you come up with a solution for ATC. 

But here we go respectfully asking you guys: Are you delivering soon ATC support, hardware/software? or is this feature just abandoned? 

Thanks a lot guys..!

**Comment by swarfer** [comment]
From my side (the post processor) coming when I get my X32 set up and can test it.
The X32 runs GRBLhal which can/does support tool changes.  
An ATC may require some custom compiling of GRBLhal or maybe just a macro inserted into the Gcode for each toolchange, this is quite easy to do in the post.

**Comment by cecheverria** [comment]
@swarfer, wishing you the best testing out on the new X32, please post what results you get on.

I'm not the master of CNC/macros, that's why I'm looking for a stack that solves complexity as much as it could.
